File: its-my-training-day/algorithms/leetcode-top-75/questions/bit-manipulation/bitwise/sum-of-two-integers/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const sumOfTwoIntegers = (a: number, b: number): number => {
  let x = a;
  let y = b;
  const mask = 0xffffffff;
  while (y !== 0) {
    const carry = (x & y) & mask;
    x = (x ^ y) & mask;
    y = (carry << 1) & mask;
  }
  return x | 0;
};

File: its-my-training-day/algorithms/leetcode-top-75/questions/bit-manipulation/bitwise/sum-of-two-integers/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def sum_of_two_integers(a: int, b: int) -> int:
    mask = 0xFFFFFFFF
    while b != 0:
        carry = (a & b) & mask
        a = (a ^ b) & mask
        b = (carry << 1) & mask
    if a & (1 << 31):
        return ~(a ^ mask)
    return a
